drivers/gpio/gpio_rt1718s_alert.c

```diff
@@ -0,0 +1,161 @@
+			 */
+			alert &= mask;
+			if (alert) {
+				/* Clear all alert bits that causes the interrupt */
+				if (rt1718s_reg_burst_write(config->rt1718s_dev[i],
+							    RT1718S_REG_ALERT, (uint8_t *)&alert,
+							    sizeof(alert))) {
+					k_sem_give(&data_rt1718s->lock_tcpci);
+					LOG_ERR("i2c access failed");
+					continue;
+				}
+			}
+			k_sem_give(&data_rt1718s->lock_tcpci);
+
+			/* There are a few sources of the vendor
+			 * defined alert for the RT1718S, but handle
+			 * only GPIO at the moment.
+			 */
+			if (alert & RT1718S_REG_ALERT_VENDOR_DEFINED_ALERT) {
+				rt1718s_gpio_alert_handler(config->rt1718s_dev[i]);
+			}
+		}
+		/* While the interrupt signal is still active, we have more work to do. */
+	} while (gpio_pin_get_dt(&config->irq_gpio));
+}
+
+static int rt1718s_alert_init(const struct device *dev)
+{
+	const struct rt1718s_alert_config *const config = dev->config;
+	struct rt1718s_alert_data *const data = dev->data;
+	int ret;
+
+	/* Check RT1718S devices are all ready. */
+	for (int i = 0; i < config->rt1718s_num; i++) {
+		if (!device_is_ready(config->rt1718s_dev[i])) {
+			LOG_ERR("%s device not ready", config->rt1718s_dev[i]->name);
+			return -ENODEV;
+		}
+	}
+
+	/* Set the alert pin for handling the interrupt */
+	k_work_init(&data->alert_worker, rt1718s_alert_worker);
+
+	if (!device_is_ready(config->irq_gpio.port)) {
+		LOG_ERR("%s device not ready", config->irq_gpio.port->name);
+		return -ENODEV;
+	}
+
+	gpio_pin_configure_dt(&config->irq_gpio, GPIO_INPUT);
+
+	gpio_init_callback(&data->gpio_cb, rt1718s_alert_callback, BIT(config->irq_gpio.pin));
+
+	ret = gpio_add_callback(config->irq_gpio.port, &data->gpio_cb);
+	if (ret < 0) {
+		return ret;
+	}
+
+	gpio_pin_interrupt_configure_dt(&config->irq_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+
+	return 0;
+}
+
+/* RT1718S alert driver must be initialized after RT1718S chip driver */
+BUILD_ASSERT(CONFIG_RT1718S_ALERT_INIT_PRIORITY > CONFIG_RT1718S_INIT_PRIORITY);
+
+#define RT1718S_DEV_AND_COMMA(node_id, prop, idx)                                         \
+	DEVICE_DT_GET(DT_PHANDLE_BY_IDX(node_id, prop, idx)),
+
+#define RT1718S_ALERT_DEVICE_INSTANCE(inst)                                               \
+	const struct device *rt1718s_dev_##inst[] = {                                     \
+		DT_INST_FOREACH_PROP_ELEM(inst, rt1718s_dev, RT1718S_DEV_AND_COMMA)};     \
+	static const struct rt1718s_alert_config rt1718s_alert_cfg_##inst = {             \
+		.irq_gpio = GPIO_DT_SPEC_INST_GET(inst, irq_gpios),                       \
+		.rt1718s_dev = &rt1718s_dev_##inst[0],                                    \
+		.rt1718s_num = DT_INST_PROP_LEN(inst, rt1718s_dev),                       \
+	};                                                                                \
+	static struct rt1718s_alert_data rt1718s_alert_data_##inst = {                    \
+		.alert_dev = DEVICE_DT_INST_GET(inst),                                    \
```

Comment:
Can this go in the `config` structure? Don't see it being changed in runtime.

Reply:
The device pointer is necessary here because I need to get the pointer in the `rt1718s_alert_worker` function, having only a pointer to the data structure.

Reply:
Oh I see, thank for explaining.